Persistence Layer for Rest API server

PhilMFischer opened the pull request:
Implements Server Repository and Persistence Management. It includes

- Repository Management
- ServerRepositories with access to backend
- Backend creation with credentials management based on VersionControlSystem

Closes #537

andriikovalov-dlr committed:
Planned classes for repositoriy persistence (Task #537)

[skip ci]

franzTobiasDLR committed:
Add basic RepoRegistry - (Task #537)

franzTobiasDLR committed:
Add loading of repo properties - (Task #537)

andriikovalov-dlr committed:
Changed repository in ServerRepository to virsat repository (Task #537)

[skip ci]

andriikovalov-dlr committed:
Convenience method for getting server property (Task #537)

[skip ci]

andriikovalov-dlr committed:
Added enum for version control system (Task #537)

[skip ci]

andriikovalov-dlr committed:
Individual property getters in ServerConfiguration (Task #537)

[skip ci]

franzTobiasDLR committed:
Add save method to repo config - (Task #537)

franzTobiasDLR committed:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of https://github.com/virtualsatellite/VirtualSatellite4-Core.git into feature/task_537-Persistence-for-server-repositories

franzTobiasDLR committed:
Ignore generated test file - (Task #537)

Commit:
Adding Project Name to Repo Config - (Task #537)

- Added field to hold user specified project name

---
Task #537: Persistency for server project repositories

Commit:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of https://github.com/virtualsatellite/VirtualSatellite4-Core.git into feature/task_537-Persistence-for-server-repositories

Commit:
Changed code for creating a new project - (Task #537)

- Moved code from wizzard to project commons
- Improved progress reporting in project creation
- Non UI code needed for creating new projects in RepoConfiguration

---
Task #537: Persistency for server project repositories

andriikovalov-dlr committed:
Removed workspace from ServerConfiguration (Task #537)

It is better to set it in standard eclipse way

[skip ci]

andriikovalov-dlr committed:
Changed RepositoryConfiguration backend from string to enum (Task #537)

[skip ci]

Commit:
Started Implementation on Repository Server - (Task #537)

- Added Implementation
- Added Skeleton for test cases
- Added dependency to project plugin

---
Task #537: Persistency for server project repositories

Commit:
Merged Manifest File - (Task #537)

- Merged dependencies

---
Task #537: Persistency for server project repositories

Commit:
Fixed test Case with Git Enum - (Task #537)

Put your Message here.

---
Task #537: Persistency for server project repositories

franzTobiasDLR committed:
Add save to server configurations - (Task #537)

franzTobiasDLR committed:
Add new test class to all tests - (Task #537)

andriikovalov-dlr committed:
Getter for repositoryConfiguration ServerRepository (Task #537)

[skip ci]

franzTobiasDLR committed:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of
https://github.com/virtualsatellite/VirtualSatellite4-Core.git into
feature/task_537-Persistence-for-server-repositories

Conflicts:
	de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/configuration/RepositoryConfigurationTest.java
	de.dlr.sc.virsat.server/src/de/dlr/sc/virsat/server/configuration/RepositoryConfiguration.java
	de.dlr.sc.virsat.server/src/de/dlr/sc/virsat/server/configuration/ServerConfiguration.java

franzTobiasDLR committed:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of https://github.com/virtualsatellite/VirtualSatellite4-Core.git into feature/task_537-Persistence-for-server-repositories

andriikovalov-dlr committed:
Test for ServerRepoHelper (Task #537)

[skip ci]

andriikovalov-dlr committed:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of
https://github.com/virtualsatellite/VirtualSatellite4-Core.git into
feature/task_537-Persistence-for-server-repositories

Conflicts:
	de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/test/AllTests.java

andriikovalov-dlr committed:
ServerRepoHelper initializes RepoRegistry (Task #537)

andriikovalov-dlr committed:
ServerRepoHelper writes RepositoryConfiguration to file (Task #537)

andriikovalov-dlr committed:
Removed redundant method in ServerRepoHelper (Task #537)

andriikovalov-dlr committed:
Initializing projects on server startup (Task #537)

Commit:
Adding test cases for ServerRepository - (Task #537)

---
Task #537: Persistency for server project repositories

Commit:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of https://github.com/virtualsatellite/VirtualSatellite4-Core.git into feature/task_537-Persistence-for-server-repositories

Commit:
Merge remote-tracking branch 'origin/development' into feature/task_537-Persistence-for-server-repositories

Commit:
Fixing location of test.properties file - (Task #537)

- Moving test file for test case into workspace of executed test
environment.

---
Task #537: Persistency for server project repositories

Commit:
Added project test dependency - (Task #537)

- Added project.test dependency to fragment if server.test

---
Task #537: Persistency for server project repositories

Commit:
Merge remote-tracking branch 'origin/development' into feature/task_537-Persistence-for-server-repositories

Commit:
Changing Version Control Backend Checkout - (Task #537)

- Path to local repo and the path of the project to load, are not
necessarily the same. Added new parameter for the path to the locally
checkout repo additionally to the project descriptor.

---
Task #537: Persistency for server project repositories

Commit:
Added Parameter to RepositoryConfiguration - (Task #537)

- Added property for localPath in Repository pointing to the project
- Typed remoteUri and File to specific objects rather than pure strings.

---
Task #537: Persistency for server project repositories

Commit:
Added Atomic Exception for handling lambdas - (Task #537)

- Added AtomicException for handling exceptions thrown over the scope of
lambdas.

---
Task #537: Persistency for server project repositories

Commit:
Refactored ServerRepository - (Task #537)

- Refactored methods for checking out project
- Refactored methods syncing the project
- Refactored method to remove the project/repository

- Tests still need to be written

---
Task #537: Persistency for server project repositories

SaMuellerDLR committed:
Externalized logic for backend creation - (Task #540)

- Moved the logic for backend creation into a in the team plugin. This
way we have access to all the svn and git dependencies without needing
to add them to the server plugin.

---
Task #540: Add REST Endpoint for Management of Repositories

SaMuellerDLR committed:
Inserted credentials into SVN remote location - (Task #537)


---
Task #537: Persistency for server project repositories

Commit:
Implementing test cases - (Task #537)

- Test Cases for ServerRepository

---
Task #537: Persistency for server project repositories

Commit:
Merging with latest changes on authentication - (Task #537)



---
Task #537: Persistency for server project repositories

Commit:
Improving test cases - (Task #537)


---
Task #537: Persistency for server project repositories

Commit:
ISsue with Repository Mapping - (Task #537)

Put your Message here.

---
Task #537: Persistency for server project repositories

SaMuellerDLR committed:
SVN Password & Username are now properly stored and tested - (Task #537)

- Added test case for testing the svn backend creation
- Properly storing the username and password now
- Refactored SVN repo creation to a serparate operation for better
reusability in test cases

---
Task #537: Persistency for server project repositories

SaMuellerDLR committed:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of https://github.com/virtualsatellite/VirtualSatellite4-Core into feature/task_537-Persistency_for_serve

SaMuellerDLR committed:
Checking if remote repo exists - (Task #537)

- Implemented check in the update functionality of the git backend to
see if a remote exists for pulling.

---
Task #537: Persistency for server project repositories

Commit:
Adding new method to backend - (Task #537)

- Added method for connecting project

---
Task #537: Persistency for server project repositories

Commit:
Merge - (Task #537)

Put your Message here.

---
Task #537: Persistency for server project repositories

SaMuellerDLR committed:
Changed SVN backend checkout to return new project - (Task #537)


---
Task #537: Persistency for server project repositories

andriikovalov-dlr committed:
Added project repositories property to ServerConfiguration (Task #537)

Commit:
Reverted changes on RepoCOnfiguration - (Task #537)

- Also moved place of creation for projects on checkout.

---
Task #537: Persistency for server project repositories

Commit:
Merging and Fixing - (Task #537)

- Fixed checkout in both svn and git backend

---
Task #537: Persistency for server project repositories

Commit:
Added comments - (Task #537)

Put your Message here.

---
Task #537: Persistency for server project repositories

Commit:
Added method to checkout or update project - (Task #537)

- Added method to update or checkout repository depending on if it
exists or not.

---
Task #537: Persistency for server project repositories

Commit:
Improved Versioning Backend Test - (Task #537)

- Improved Checkout test to now checkout project into a subfolder for
local repos, rather than checking it out into the repoHome root
location.

---
Task #537: Persistency for server project repositories

Commit:
Implemented method for updateOrCheckout - (Task #537)

- Added test case
- Improved commit messages

---
Task #537: Persistency for server project repositories

Commit:
Updated test case for retrieveEd... - (Task #537)

- generally improved the test case

---
Task #537: Persistency for server project repositories

Commit:
implemented test case checking out 2 projects - (Task #537)

- Added test case to checkout two projects from one repository
- Minor fixes to Server Repository to improve sub folder generation

---
Task #537: Persistency for server project repositories

Commit:
Removing unneeded commits in git - (Task #537)

---
Task #537: Persistency for server project repositories

Commit:
Fixed all testcases - (Task #537)

---
Task #537: Persistency for server project repositories

Commit:
Fixing test for maven tycho - (Task #537)

- added better cleanup for RepoHelperTest on registry singleton instance

---
Task #537: Persistency for server project repositories

Commit:
Fixed checkstyle - (Task #537)

- Removed unnecessary import

---
Task #537: Persistency for server project repositories

SaMuellerDLR committed:
FIxed incomplete assertion - (Task #537)


---
Task #537: Persistency for server project repositories

SaMuellerDLR committed:
Merge branch 'development' into feature/task_537-Persistency_for_serve

Conflicts:
	de.dlr.sc.virsat.team/src/de/dlr/sc/virsat/team/git/VirSatGitVersionControlBackend.java
	de.dlr.sc.virsat.team/src/de/dlr/sc/virsat/team/svn/VirSatSvnVersionControlBackend.java

PhilMFischer commented on de.dlr.sc.virsat.server/src/de/dlr/sc/virsat/server/repository/ServerRepoHelper.java:
```diff
@@ -0,0 +1,81 @@
+	 * @throws FileNotFoundException 
+	 * @throws URISyntaxException 
+	 */
+	private static void registerRepositoryConfiguration(Path repositoryConfigurationFile) throws FileNotFoundException, IOException, URISyntaxException {
```
Well spotted. Removed the Exception as suggested.

Commit:
Changes due to review - (Task #537)

- removed redundant exception in throw statement of method

---
Task #537: Persistency for server project repositories

Commit:
Minor fixes due to review - (Task #537)

- Had to add a remote repo to also test ramping up a SVN backend in the
repo registry tests.

---
Task #537: Persistency for server project repositories

PhilMFischer commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/repository/ServerRepoHelperTest.java:
```diff
@@ -0,0 +1,105 @@
+	public void testLoadRepositoryConfigurations() throws IOException {
+		String svnProjectName = "GitProject1";
+		String gitProjectName = "GitProject2";
+		createTempRepoConfigFile(configsDir, svnProjectName, VersionControlSystem.GIT);
```
Fixed, a remote SVN repo had to be created first, otherwise the instantiation of an SVN Repo and Server will fail.

PhilMFischer commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/repository/ServerRepoHelperTest.java:
```diff
@@ -0,0 +1,105 @@
+		ServerRepoHelper.initRepoRegistry();
+		
+		assertEquals(2, repoRegistry.getRepositories().size());
+		assertEquals(VersionControlSystem.GIT, repoRegistry.getRepository(svnProjectName).getRepositoryConfiguration().getBackend());
```
See above.

Commit:
Fixed dependency to CreateSvnServerOperation - (Task #537)

- Tycho cannot resolve fragments as Eclispe IDE does, therefore classes
which are reused in another test fragment need to be placed in a package
which does not exist in the host and this package needs to be implicitly
imported to the using fragment.

---
Task #537: Persistency for server project repositories

Commit:
Fix location of test resource - (Task #537)

Moving test reosurce from plugin path to runtime workspace

---
Task #537: Persistency for server project repositories

andriikovalov-dlr commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/repository/ServerRepositoryTest.java:
```diff
@@ -0,0 +1,241 @@
+	
+		RevCommit logAfterSync = Git.open(pathRepoRemote.toFile()).log().call().iterator().next();
+		
+		assertThat("remote repo does not yet have a commit as expected", logAfterSync.getFullMessage(), containsString(""));
```
Commit message contains empty string?

andriikovalov-dlr reviewed (changes requested):
Some minor remarks

SaMuellerDLR committed:
Removed unused import - (Task #537)

- Removed unused import that was somehow not marked as unused in eclipse

---
Task #537: Persistency for server project repositories

Commit:
Removed bad import - (Task #537)

- Warning on travis about uneeded import

---
Task #537: Persistency for server project repositories

Commit:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of https://github.com/virtualsatellite/VirtualSatellite4-Core.git into feature/task_537-Persistence-for-server-repositories

Commit:
Update de.dlr.sc.virsat.commons.test/src/de/dlr/sc/virsat/commons/exception/AtomicExceptionTest.java

Co-Authored-By: andriikovalov-dlr <48556822+andriikovalov-dlr@users.noreply.github.com>

PhilMFischer commented on de.dlr.sc.virsat.server.test/META-INF/MANIFEST.MF:
```diff
@@ -6,6 +6,10 @@ Bundle-Version: 4.11.0.qualifier
+ org.eclipse.team.svn.core
 Export-Package: de.dlr.sc.virsat.server.test
 Automatic-Module-Name: de.dlr.sc.virsat.server.test
+Import-Package: de.dlr.sc.virsat.project.test
```
This is because Tycho maven resolves fragments differently then an eclipse application. If you access a class from one fragment into another fragment, it needs to be imported implicitly via the IMPORT-PACKAGE. Otherwise the maven/tycho will run into a compilation issue.

PhilMFischer commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/configuration/RepositoryConfigurationTest.java:
```diff
@@ -0,0 +1,125 @@
+	@Test
+	public void testSaveProperties() throws IOException, URISyntaxException {
+		
+		File wsRootFile = ResourcesPlugin.getWorkspace().getRoot().getRawLocation().toFile();
```
Hmm yes sounds like an idea.

PhilMFischer commented on de.dlr.sc.virsat.server/resources/server.properties:
```diff
@@ -1,2 +1,5 @@
 # Default properties for Virtual Satellite Server configuration
-config.area = /resources
\ No newline at end of file
+repository.configurations.dir = ~/virtual_satellite/configurations/repositories
```
I have seen this quite often recently. I have no idea what going on here.

PhilMFischer commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/configuration/ServerConfigurationTest.java:
```diff
@@ -41,7 +50,24 @@ public void testLoadProperties() throws IOException {
+	@Test
+	public void testSaveProperties() throws IOException {
+		
+		final String TEST_FILE_NAME = "test.properties";
```
Agreee

Commit:
Update de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/repository/RepoRegistryTest.java

Co-Authored-By: andriikovalov-dlr <48556822+andriikovalov-dlr@users.noreply.github.com>

Commit:
Update de.dlr.sc.virsat.team/src/de/dlr/sc/virsat/team/IVirSatVersionControlBackend.java

Co-Authored-By: andriikovalov-dlr <48556822+andriikovalov-dlr@users.noreply.github.com>

Commit:
Update de.dlr.sc.virsat.team/src/de/dlr/sc/virsat/team/IVirSatVersionControlBackend.java

Co-Authored-By: andriikovalov-dlr <48556822+andriikovalov-dlr@users.noreply.github.com>

Commit:
Changed from WS to temp file in test - (Task #537)

- Creating .properties file in temporary folder instead of the eclipse
runtime workspace.

---
Task #537: Persistency for server project repositories

Commit:
Converted line delimiters to windows style - (Task #537)

- fixed line delimiters in server properties

---
Task #537: Persistency for server project repositories

PhilMFischer commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/configuration/ServerConfigurationTest.java:
```diff
@@ -11,24 +11,33 @@
 
 public class ServerConfigurationTest {
+	
+	public static final String REPOSITORY_CONFIGURATIONS_DIR = "some/dir";
```
Done

Commit:
Improved Test Case for Repo Registry - (Task #537)

- Added further asserts as stated in Review

---
Task #537: Persistency for server project repositories

PhilMFischer commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/repository/RepoRegistryTest.java:
```diff
@@ -0,0 +1,51 @@
+		
+		RepoRegistry.getInstance().getRepositories().put(REPO_NAME, TEST_REPO);
+		
+		ServerRepository returnedFromRegistry = RepoRegistry.getInstance().getRepositories().get(REPO_NAME);
```
Done

PhilMFischer commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/repository/RepoRegistryTest.java:
```diff
@@ -0,0 +1,51 @@
+		
+		//Create basic test repsoitory 
+		final ServerRepository TEST_REPO = new ServerRepository(new File(""), new RepositoryConfiguration("", "", "", VersionControlSystem.GIT, "", ""));
+		
```
Done

Commit:
Fixes due to review - (Task #537)

- Using setter of ServerCOnfiguration rather than writing the
properties.

---
Task #537: Persistency for server project repositories

PhilMFischer commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/repository/ServerRepoHelperTest.java:
```diff
@@ -0,0 +1,137 @@
+		configsDir = Files.createTempDirectory("test_repo_configs");
+		
+		// Overwrite path to repo config files
+		ServerConfiguration.getProperties().setProperty(ServerConfiguration.REPOSITORY_CONFIGURATIONS_DIR_KEY, configsDir.toString());
```
I agree, wasn't it you coding that?

andriikovalov-dlr replied:
Yes it was :) I think we made getters later

PhilMFischer commented on de.dlr.sc.virsat.server.test/src/de/dlr/sc/virsat/server/repository/ServerRepoHelperTest.java:
```diff
@@ -0,0 +1,137 @@
+
+		RepoRegistry.getInstance().getRepositories().clear();
+		
+		svnPathRepoRemote = Files.createTempDirectory("VirtualSatelliteSvnRemote_");
```
I did, got complaints from @JAmmermann-DLR . Anyway test coverage is now increase, so we leave it with svn.

PhilMFischer commented on de.dlr.sc.virsat.server/src/de/dlr/sc/virsat/server/VirSatServerApplication.java:
```diff
@@ -46,7 +78,8 @@ public void stop() {
 			e.printStackTrace();
-		}		
+		}
+
```
Can't find it in the code.

Commit:
Update de.dlr.sc.virsat.server/src/de/dlr/sc/virsat/server/configuration/RepositoryConfiguration.java

Co-Authored-By: andriikovalov-dlr <48556822+andriikovalov-dlr@users.noreply.github.com>

Commit:
Removed try with resource from Helper - (Task #537)

- Removed try with resource since it does not need to close files.

---
Task #537: Persistency for server project repositories

PhilMFischer commented on de.dlr.sc.virsat.server/src/de/dlr/sc/virsat/server/repository/ServerRepoHelper.java:
```diff
@@ -0,0 +1,81 @@
+						registerRepositoryConfiguration(t);
+					} catch (IOException e) {
+						throw new UncheckedIOException(e);
+					} catch (URISyntaxException e) {
```
URI Exception is needed, but removed the try with resource.

PhilMFischer commented on de.dlr.sc.virsat.server/src/de/dlr/sc/virsat/server/repository/ServerRepoHelper.java:
```diff
@@ -0,0 +1,81 @@
+	 * @throws FileNotFoundException 
+	 * @throws URISyntaxException 
+	 */
+	private static void registerRepositoryConfiguration(Path repositoryConfigurationFile) throws IOException, URISyntaxException {
```
checked is thrown

Commit:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of https://github.com/virtualsatellite/VirtualSatellite4-Core.git into feature/task_537-Persistence-for-server-repositories

PhilMFischer commented on de.dlr.sc.virsat.server/src/de/dlr/sc/virsat/server/repository/ServerRepository.java:
```diff
@@ -0,0 +1,246 @@
+	private VirSatTransactionalEditingDomain ed;
+	private IVirSatVersionControlBackend versionControlBackEnd;
+	@SuppressWarnings("unused")
+	private File localRepositoryHome;
```
Removed it, there was an initial idea, but not needed anymore.

PhilMFischer commented on de.dlr.sc.virsat.server/src/de/dlr/sc/virsat/server/repository/ServerRepository.java:
```diff
@@ -0,0 +1,246 @@
+		atomicException.throwIfSet();
+	}
+	
+	public static final String SERVER_REPOSITORY_COMMIT_PUSH_MESSAGE = "Server Local Commit Before Push: ";
```
I don't agree. Having this message public allows others to access and search for this message in their code.

Commit:
MInor changes due to review - (Task #537)

---
Task #537: Persistency for server project repositories

andriikovalov-dlr committed:
Fixed commit message assertion in ServerRepositoryTest (Task #537)

andriikovalov-dlr committed:
Moved a constant to the top of the class (Task #537)

Commit:
Fixing import warnings  - (Task #537)

- Removed non needed imports for AtomicExceptionTest

---
Task #537: Persistency for server project repositories

Commit:
Merge branch 'feature/task_537-Persistence-for-server-repositories' of https://github.com/virtualsatellite/VirtualSatellite4-Core.git into feature/task_537-Persistence-for-server-repositories
